src/runtime_src/core/include/experimental/xrt_queue.h

```diff
@@ -0,0 +1,297 @@
+    {}
+
+    event(const event& rhs) : m_content(rhs.m_content)
+    {}
+
+    // event() - event constructor for std::shared_future
+    //
+    // @e : shared_future to type erase
+    template <typename ValueType>
+    event(std::shared_future<ValueType> e)
+      : m_content(new event_holder<ValueType>(std::move(e)))
+    {}
+
+    event&
+    operator=(event&& rhs)
+    {
+      m_content = std::move(rhs.m_content);
+      return *this;
+    }
+
+    operator bool() const
+    {
+      return m_content!=nullptr;
+    }
+
+    void
+    wait() const
+    {
+      if (m_content)
+        m_content->wait();
+    }
+  };
+
+private:
+  // Add task to queue
+  void
+  add_task(task&& ev);
+
+public:
+  /**
+   * queue() - Constructor for queue object
+   *
+   * The queue is constructed with one consumer thread.
+   */
+  queue();
+
+  /**
+   * enqueue() - Enqueue a callable
+   *
+   * @param c
+   *   Callable function, typically a lambda
+   * @return
+   *   Future result of the function (std::future)
+   *
+   * A callable is an argument less lambda function.  The function is
+   * executed asynchronously by the queue consumer (worker thread)
+   * once all previous enqueued operations have completed.
+   *
+   * Upon completion the returned future becomes valid and will
+   * contain the return value of executing the lambda.
+   */
+  template <typename Callable>
+  auto
+  enqueue(Callable&& c)
+  {
+    using value_type = decltype(c());
+    std::packaged_task<value_type()> task{[cc = std::move(c)] { return cc(); }};
+    std::shared_future<value_type> f{task.get_future()};
+    add_task(std::move(task));
+    return f;
+  }
+
+  /**
+   * enqueue() - Enqueue the future of an enqueued operation
+   *
+   * @param sf
+   *   The future result to wait on (std::shared_future)
+   * @return
+   *   Future of the future (std::shared_future<void>)
+   *
+   * Subsequent enqueued task blocks until the enqueued future is
+   * valid.
+   *
+   * This type of enqueued future is used for synchronization between
+   * multiple queues.
+   */
+  template <typename ValueType>
+  auto
+  enqueue(std::shared_future<ValueType> sf)
+  {
+    std::packaged_task<void()> task([evc = xrt::queue::event{std::move(sf)}]() { evc.wait(); });
+    std::shared_future<void> f{task.get_future()};
+    add_task(std::move(task));
+    return f;
+  }
+
+  /**
+   * enqueue() - Enqueue an event (type erased future)
+   *
+   * @oaram ev
+   *   Event to enqueue
+   * @param
+   *   Future of event (std::shared_future<void>)
+   *
+   * Subsequent enqueued task blocks until the enqueued event is
+   * valid.
+   *
+   * This type of enqueued event is used for synchronization between
+   * multiple queues.
+   */
+  auto
+  enqueue(xrt::queue::event ev)
+  {
+    std::packaged_task<void()> task([evc = xrt::queue::event{std::move(ev)}]() { evc.wait(); });
+    std::shared_future<void> f{task.get_future()};
+    add_task(std::move(task));
+    return f;
```

Comment:
Globally you could just rewrite the 2 previous `enqueue` by just calling the first generic enqueue on the lambda.

Reply:
Yeah